File: Coding_Challenges/CC72/cc72.php
<?php
include("../../translationutils/jsarray.php");
$height = 500;
$width = 500;
class Point{
  function __construct(){
    global $height;
    global $width;
    $this->x = rand(0,$width);
    $this->y = rand(0,$height);
    if($this->x > $this->y){
      $this->label = 1;
    } else {
      $this->label = -1;
    }
  }
}
// The activation function
function sign($n){
  if($n >= 0){
    return 1;
  } else {
    return -1;
  }
}
class Perceptron{
  // Constructor
  function __construct(){
    // Initialize the weights randomly
    $this->weights = array(rand(-1000,1000)/1000,rand(-1000,1000)/1000);
    $this->lr = 0.1;
  }
  function guess($inputs){
    $sum = 0;
    foreach(array_combine($this->weights,$inputs) as $weight => $input){
      $sum += $input*$weight;
    }
    return sign($sum);
  }
  function train($inputs,$target){
    $guess = sign($this->guess($inputs));
    $error = $target - $guess;
    foreach(array_combine($this->weights,$inputs) as $weight => $input){
      $weight += $error * $input * $this->lr;
    }
  }
}

// Main Code
$brain = new Perceptron();
$points = new jsarray();
for($i = 0;$i<10;$i++){
  $points->push(new Point());
}
$trainingindex = 0;
for($i = 0;$i<1;$i++){
  $training = $points($trainingindex,NULL);
  $inputs = new jsarray($training->x,$training->y);
  $target = $training->label;
  $brain->train($inputs->vals,$target);
  $trainingindex++;
  if($trainingindex == $points->length) {
    $trainingindex = 0;
  }
}
?>

File: Coding_Challenges/CC72/cc72.java
class Point {
  float x;
  float y;
  int label;

  Point() {
    x = random(width);
    y = random(height);

    if (x > y) {
      label = 1;
    } else {
      label = -1;
    }
  }

  void show() {
    stroke(0);
    if (label == 1) {
      fill(255);
    } else {
      fill(0);
    }
    ellipse(x, y, 32, 32);
  }
}
// The activation function
int sign(float n) {
  if (n >= 0) {
    return 1;
  } else {
    return -1;
  }
}


class Perceptron {
  float[] weights = new float[2];
  float lr = 0.1;

  // Constructor
  Perceptron() {
    // Initialize the weights randomly
    for (int i = 0; i < weights.length; i++) {
      weights[i] = random(-1, 1);
    }
  }

  int guess(float[] inputs) {
    float sum = 0;
    for (int i = 0; i < weights.length; i++) {
      sum += inputs[i]*weights[i];
    }
    int output = sign(sum);
    return output;
  }

  void train(float[] inputs, int target) {
    int guess = guess(inputs);
    int error = target - guess;
    // Tune all the weights
    for (int i = 0; i < weights.length; i++) {
      weights[i] += error * inputs[i] * lr;
    }
  }
}

Perceptron brain;

Point[] points = new Point[100];

int trainingIndex = 0;

void setup() {
  size(800, 800);
  brain = new Perceptron();

  for (int i = 0; i < points.length; i++) {
    points[i] = new Point();
  }



  float[] inputs = {-1, 0.5};
  int guess = brain.guess(inputs);
  println(guess);
}

void draw() {
  background(255);
  stroke(0);
  line(0, 0, width, height);
  for (Point pt : points) {
    pt.show();
  }

  for (Point pt : points) {
    float[] inputs = {pt.x, pt.y};
    int target = pt.label;
    int guess = brain.guess(inputs);
    if (guess == target) {
      fill(0, 255, 0);
    } else {
      fill(255, 0, 0);
    }
    noStroke();
    ellipse(pt.x, pt.y, 16, 16);
  }

  Point training = points[trainingIndex];
  float[] inputs = {training.x, training.y};
  int target =training.label;
  brain.train(inputs, target);
  trainingIndex++;
  if (trainingIndex == points.length) {
    trainingIndex = 0;
  }
}

void mousePressed() {
  for (Point pt : points) {
  }
}
